site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,993 @@
+    {
+      "root": {
+        "names": [
+          "product_name",
+          "product_id",
+          "sales"
+        ],
+        "input": { ... }
+      }
+    }
+  ]
+}
+```
+
+The `relations` field is a list of Root relations. Most queries only have one 
+root relation, but the spec allows for multiple so a common plan could be
+referenced by other plans, sort of like a CTE (Common Table Expression) from SQL.
+The root relation provides the final column names for our query. The input to
+this relation is our aggregate relation (which contains all the other relations
+as children).
+
+For extensions, we need to provide `extensionUris` with the locations of the 
+YAML files we used and `extensions` with the list of functions we used and which
+extension they come from.
+
+In our query, we used:
+
+ * `index_in` (0), from `functions_string.yaml`,
+ * `is_null` (1), from `functions_comparison.yaml`,
+ * `equal` (2), from `functions_comparison.yaml`,
+ * `sum` (3), from `functions_arithmetic_decimal.yaml`,
+ * `multiply` (4), from `functions_arithmetic_decimal.yaml`.
+
+So first we can create the three extension uris:
+
+```json
+[
+  {
+    "extensionUriAnchor": 0,
+    "uri": "https://github.com/substrait-io/substrait/blob/main/extensions/functions_string.yaml"
+  },
+  {
+    "extensionUriAnchor": 1,
+    "uri": "https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml"
+  },
+  {
+    "extensionUriAnchor": 2,
+    "uri": "https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic_decimal.yaml"
+  }
+]
+```
+
+Then we can create the extensions:
+
+```json
+[
+  {
+    "extensionFunction": {
+      "extensionUriReference": 0,
+      "functionAnchor": 0,
+      "name": "index_in"
+    }
+  },
+  {
+    "extensionFunction": {
+      "extensionUriReference": 1,
+      "functionAnchor": 1,
+      "name": "is_null"
+    }
+  },
+  {
+    "extensionFunction": {
+      "extensionUriReference": 1,
+      "functionAnchor": 2,
+      "name": "equal"
+    }
+  },
+  {
+    "extensionFunction": {
+      "extensionUriReference": 2,
+      "functionAnchor": 3,
+      "name": "sum"
+    }
+  },
+  {
+    "extensionFunction": {
+      "extensionUriReference": 2,
+      "functionAnchor": 4,
+      "name": "multiply"
+    }
+  }
+]
+```
+
+Our plan outputted in full is: [final_plan.json](./final_plan.json).
+
+## Next steps
+
+**Validate and introspect plans using 
+[substrait-validator](https://github.com/substrait-io/substrait-validator)**.
+Amongst other things, this tool can show what the current schema and column
+indices are at each point in the plan. Try downloading the final plan JSON above
+and generating an HTML report on the plan with:
+
+```sh
+substrait-validator final_plan.json --out-file output.html
+```
+
+**Generate more Substrait plans from SQL using DuckDB's compiler**. See the 
+documentation on
+[the DuckDB Substrait extension](https://duckdb.org/docs/extensions/substrait).
```

Comment:
I'd love to point readers to one of the tools that currently can be used to generate Substrait from SQL.

DuckDB looked easy to install, but not sure if they want that attention yet. I noticed the instructions on the website had some typos and the provided example errors when run in Python. From the github repo, they aren't accepting bug reports except from those who have pre-arranged agreements.

I didn't see a user-facing way to install Isthmus (just compilation instructions), but correct me if I'm wrong.

Reply:
Frankly I also think the existing docs could be improved further. I am not an expert in Graal based builds, but I also think it would be great to document the build, install and usage of isthmus with little more detail. I guess non-Java developers would also use this.